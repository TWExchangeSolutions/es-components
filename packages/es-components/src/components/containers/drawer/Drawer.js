import React, { Children } from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';
import { noop } from 'lodash';
import uncontrollable from 'uncontrollable';
import classnames from 'classnames';

import DrawerPanel from './DrawerPanel';

function drawerPanelPropType(props, propName, componentName) {
  const value = props[propName];
  const errMsg = `${componentName} ${propName} contains an element that is not a Drawer.Panel.`;

  if (Array.isArray(value) && value.some(({ type }) => type !== DrawerPanel)) {
    return new Error(errMsg);
  }
  if (!Array.isArray(value) && value.type !== DrawerPanel) {
    return new Error(errMsg);
  }
  return null;
}

const StyledDrawer = styled.div`
  background-color: ${props => props.theme.colors.white};
  border-top: 1px solid ${props => props.theme.colors.gray3};
  box-shadow: 0 1px 1px ${props => props.theme.colors.boxShadowLight};
  margin-bottom: 25px;
`;

export const Drawer = props => {
  const {
    activeKeys,
    children,
    className,
    closedIconName,
    isAccordion,
    onActiveKeysChanged,
    openedIconName
  } = props;

  const onItemClick = key => {
    let nextActiveKeys = [...activeKeys];

    if (isAccordion) {
      nextActiveKeys = nextActiveKeys[0] === key ? [] : [key];
    } else {
      const index = nextActiveKeys.indexOf(key);
      const isOpen = index > -1;

      if (isOpen) {
        nextActiveKeys.splice(index, 1);
      } else {
        nextActiveKeys.push(key);
      }
    }

    onActiveKeysChanged(nextActiveKeys);
  };

  const getPanels = () =>
    Children.map(children, (child, index) => {
      // If there is no key provided, use the panel order as default key
      const key = child.key || String(index + 1);
      const { title, titleAside } = child.props;
      const noPadding = child.props.noPadding || false;

      let isOpen = false;
      if (isAccordion) {
        isOpen = activeKeys[0] === key;
      } else {
        isOpen = activeKeys.indexOf(key) > -1;
      }

      const childProps = {
        key,
        title,
        titleAside,
        noPadding,
        isOpen,
        children: child.props.children,
        onItemClick: () => onItemClick(key),
        closedIconName,
        openedIconName
      };

      return React.cloneElement(child, childProps);
    });

  return (
    <StyledDrawer className={classnames('es-drawer', className)}>
      {getPanels()}
    </StyledDrawer>
  );
};

Drawer.propTypes = {
  /** Set which panels are open */
  activeKeys: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  /** Should only contain one or more Drawer.Panel elements */
  children: drawerPanelPropType,
  /** Add additional CSS classes to the root drawer element */
  className: PropTypes.string,
  /** Override the default plus icon with another OE icon name */
  closedIconName: PropTypes.string,
  /** Used in uncontrolled mode to set initial drawer state */
  defaultActiveKeys: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  /** Only allows one DrawerPanel to be open at a time */
  isAccordion: PropTypes.bool,
  /** Function called when changing active keys */
  onActiveKeysChanged: PropTypes.func,
  /** Override the default minus icon with another OE icon name */
  openedIconName: PropTypes.string
};

Drawer.defaultProps = {
  activeKeys: [],
  isAccordion: false,
  closedIconName: 'add',
  onActiveKeysChanged: noop,
  openedIconName: 'minus'
};

const UncontrolledDrawer = uncontrollable(Drawer, {
  activeKeys: 'onActiveKeysChanged'
});

UncontrolledDrawer.Panel = DrawerPanel;

export default withTheme(UncontrolledDrawer);
