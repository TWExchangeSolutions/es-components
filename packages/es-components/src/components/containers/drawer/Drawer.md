Use `Drawer` and `Drawer.Panel` components to wrap content in collapsable panels. This component can be used in a controlled or uncontrolled way. When uncontrolled, you can specify `defaultActiveKeys` to set initial state. When controlled, you must supply both `activeKeys` and an `onActiveKeysChanged` handler.

### Basic Example (Uncontrolled)
```
<Drawer defaultActiveKeys={"1"}>
  <Drawer.Panel title="Panel 1">
    <p>Pretty much any content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 2">
    <p>More content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </Drawer.Panel>
</Drawer>
```


### Accordion (Uncontrolled)
Add the `isAccordion` property to change the default behavior to an accordion
style, where only one panel can be opened at a time.
```
<Drawer isAccordion>
  <Drawer.Panel title="Accordion Panel 1">
    <p>Pretty much any content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Accordion Panel 2" noPadding>
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </Drawer.Panel>
  <Drawer.Panel title="Accordion Panel 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </Drawer.Panel>
</Drawer>
```


### Advanced Example (Controlled)
Use the `noPadding` property to remove default padding within a panel (useful for lists and tables).
Use the `titleAside` property to display text or other content on the right side of the panel header.
You can customize the `key` property of each Panel; if not specified a default key value will be assigned
matching the Panel's numeric position.
Both the Drawer and Drawer.Panel components will accept additional classNames.
```
class DrawerExample extends React.Component {
  constructor() {
    this.state = { activeKeys: ["second","3"] };
    this.onActiveKeysChanged = this.onActiveKeysChanged.bind(this);
  }

  onActiveKeysChanged(value) {
    this.setState({ activeKeys: value });
  };

  render() {
    return (
      <Drawer activeKeys={this.state.activeKeys} onActiveKeysChanged={this.onActiveKeysChanged} closedIconName="hand-right" openedIconName="hand-down" className="drawer-big">
        <Drawer.Panel title="Panel One" key="first" noPadding titleAside={<em>aside text</em>}>
          <p>The default padding has been removed from this panel.</p>
        </Drawer.Panel>
        <Drawer.Panel title="Panel Two" key="second" noPadding titleAside={<Icon name='tag' />}>
          <ul>
            <li>List Item</li>
            <li>Another Item</li>
          </ul>
        </Drawer.Panel>
        <Drawer.Panel title="Panel Three" className="alt-color">
          <div>
            More content. Anim pariatur cliche reprehenderit, enim eiusmod
            high life accusamus terry richardson ad squid.
          </div>
        </Drawer.Panel>
      </Drawer>
    )
  }
}
<DrawerExample />
```
