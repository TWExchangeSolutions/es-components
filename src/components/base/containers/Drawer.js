import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';
import DrawerPanel from './DrawerPanel';
import { colors } from '../../theme';
import styled from 'styled-components';

function toArray(activeKeys) {
  let currentActiveKeys = activeKeys;
  if (!Array.isArray(currentActiveKeys)) {
    currentActiveKeys = currentActiveKeys ? [currentActiveKeys] : [];
  }
  return currentActiveKeys;
}

function drawerPanelPropType(props, propName, componentName) {
  const value = props[propName];
  const errMsg = `${componentName} ${propName} contains an element that is not a DrawerPanel.`;

  if (Array.isArray(value) && value.some(({ type }) => type !== DrawerPanel)) {
    return new Error(errMsg);
  }
  if (!Array.isArray(value) && value.type !== DrawerPanel) {
    return new Error(errMsg);
  }
  return null;
}

const StyledDrawer = styled.div`
  background-color: ${colors.white};
  border-top: 1px solid ${colors.grayLight};
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.05);
  margin-bottom: 25px;
`;

class Drawer extends Component {
  constructor(props) {
    super(props);

    const currentActiveKeys = this.props.defaultActiveKeys;
    this.state = {
      activeKeys: toArray(currentActiveKeys)
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('activeKeys' in nextProps) {
      this.setState({
        activeKeys: toArray(nextProps.activeKeys)
      });
    }
  }

  onClickItem(key) {
    return () => {
      let activeKeys = [...this.state.activeKeys];

      if (this.props.isAccordion) {
        activeKeys = activeKeys[0] === key ? [] : [key];
      } else {
        const index = activeKeys.indexOf(key);
        const isActive = index > -1;

        if (isActive) {
          activeKeys.splice(index, 1);
        } else {
          activeKeys.push(key);
        }
      }

      this.setState({ activeKeys });
    };
  }

  getPanels() {
    const activeKeys = this.state.activeKeys;

    return Children.map(this.props.children, (child, index) => {
      // If there is no key provided, use the panel order as default key
      const key = child.key || String(index);
      const header = child.props.header;
      const noPadding = child.props.noPadding || false;

      let isActive = false;
      if (this.props.isAccordion) {
        isActive = activeKeys[0] === key;
      } else {
        isActive = activeKeys.indexOf(key) > -1;
      }

      const props = {
        key,
        header,
        noPadding,
        isActive,
        children: child.props.children,
        onItemClick: this.onClickItem(key).bind(this),
        closedIconName: this.props.closedIconName,
        openedIconName: this.props.openedIconName
      };

      return React.cloneElement(child, props);
    });
  }

  render() {
    return (
      <StyledDrawer className={this.props.className}>
        {this.getPanels()}
      </StyledDrawer>
    );
  }
}

Drawer.propTypes = {
  /** Should only contain one or more DrawerPanel elements */
  children: drawerPanelPropType,
  /** Add additional CSS classes to the root drawer element */
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object
  ]),
  /** Override the default plus icon with another OE icon name */
  closedIconName: PropTypes.string,
  /** Specify which panels are opened by default */
  defaultActiveKeys: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  /** Only allows one DrawerPanel to be open at a time */
  isAccordion: PropTypes.bool,
  /** Override the default minus icon with another OE icon name */
  openedIconName: PropTypes.string
};

Drawer.defaultProps = {
  isAccordion: false,
  closedIconName: 'plus',
  openedIconName: 'minus'
};

export default Drawer;
