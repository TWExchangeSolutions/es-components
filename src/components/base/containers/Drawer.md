Use ``Drawer`` and ``DrawerPanel`` components to wrap content in collapsable panels.

### Basic Example:
```
<Drawer>
  <DrawerPanel header="Panel 1">
    <p>Pretty much any content you want in here.</p>
  </DrawerPanel>
  <DrawerPanel header="Panel 2">
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </DrawerPanel>
  <DrawerPanel header="Panel 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </DrawerPanel>
</Drawer>
```

### Advanced Example:
```
<Drawer isAccordion defaultActiveKeys="1" closedIconName="hand-right" openedIconName="hand-down" className="drawer-big">
  <DrawerPanel header="Panel 1" key="1">
    <p>Pretty much any content you want in here.</p>
  </DrawerPanel>
  <DrawerPanel header="Panel 2" key="2" noPadding>
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </DrawerPanel>
  <DrawerPanel header="Panel 3" key="3" className="alt-color">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </DrawerPanel>
</Drawer>
```
