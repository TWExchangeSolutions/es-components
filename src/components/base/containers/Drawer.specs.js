/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Drawer from './Drawer';

jest.mock('../../util/generateAlphaName', () => () => 'abcdef');

describe('drawer component', () => {
  describe('drawer', () => {
    let instanceToRender;

    beforeEach(() => {
      instanceToRender = (
        <Drawer className="important">
          <Drawer.Panel title="collapse 1" key="1" className="first" titleAside="side text">first</Drawer.Panel>
          <Drawer.Panel title="collapse 2" key="2" className="second" noPadding>second</Drawer.Panel>
          <Drawer.Panel title="collapse 3" key="3" className="third">third</Drawer.Panel>
        </Drawer>
      );
    });

    it('renders as expected', () => {
      const tree = renderer.create(instanceToRender).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('click should toggle panel state', () => {
      const drawerInstance = mount(instanceToRender);
      const firstPanel = drawerInstance.find('.first');
      const panelTitle = firstPanel.find({ role: 'tab' });

      expect(panelTitle.prop('aria-expanded')).toBe(false);

      panelTitle.simulate('click');
      expect(panelTitle.prop('aria-expanded')).toBe(true);

      panelTitle.simulate('click');
      expect(panelTitle.prop('aria-expanded')).toBe(false);
    });

    it('should allow more than one drawer to be opened at a time', () => {
      const drawerInstance = mount(instanceToRender);
      const firstPanel = drawerInstance.find('.first');
      const firstTitle = firstPanel.find({ role: 'tab' });
      const thirdPanel = drawerInstance.find('.third');
      const thirdTitle = thirdPanel.find({ role: 'tab' });

      expect(firstTitle.prop('aria-expanded')).toBe(false);
      expect(thirdTitle.prop('aria-expanded')).toBe(false);

      firstTitle.simulate('click');
      expect(firstTitle.prop('aria-expanded')).toBe(true);
      expect(thirdTitle.prop('aria-expanded')).toBe(false);

      thirdTitle.simulate('click');
      expect(firstTitle.prop('aria-expanded')).toBe(true);
      expect(thirdTitle.prop('aria-expanded')).toBe(true);
    });

    it('should set multiple default panels open at start', () => {
      const drawerInstance = mount(<Drawer defaultActiveKeys={['2', '3']}>
        <Drawer.Panel title="collapse 1" key="1">first</Drawer.Panel>
        <Drawer.Panel title="collapse 2" key="2">second</Drawer.Panel>
        <Drawer.Panel title="collapse 3" key="3">third</Drawer.Panel>
      </Drawer>);

      const panels = drawerInstance.find('[aria-expanded]');
      expect(panels.at(0).prop('aria-expanded')).toBe(false);
      expect(panels.at(1).prop('aria-expanded')).toBe(true);
      expect(panels.at(2).prop('aria-expanded')).toBe(true);
    });
  });

  describe('accordion', () => {
    let instanceToRender;

    beforeEach(() => {
      instanceToRender = (
        <Drawer isAccordion defaultActiveKeys="1">
          <Drawer.Panel title="collapse 1" key="1" className="first">first</Drawer.Panel>
          <Drawer.Panel title="collapse 2" key="2" className="second">second</Drawer.Panel>
          <Drawer.Panel title="collapse 3" key="3" className="third">third</Drawer.Panel>
        </Drawer>
      );
    });

    it('renders as expected', () => {
      const tree = renderer.create(instanceToRender).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('default active key should set panel open', () => {
      const accordionInstance = mount(instanceToRender);
      const panels = accordionInstance.find('[aria-expanded]');

      expect(panels.at(0).prop('aria-expanded')).toBe(true);
      expect(panels.at(1).prop('aria-expanded')).toBe(false);
      expect(panels.at(2).prop('aria-expanded')).toBe(false);
    });

    it('should only allow one drawer to be opened at a time', () => {
      const accordionInstance = mount(instanceToRender);
      const firstPanel = accordionInstance.find('.first');
      const firstTitle = firstPanel.find({ role: 'tab' });
      const thirdPanel = accordionInstance.find('.third');
      const thirdTitle = thirdPanel.find({ role: 'tab' });

      expect(firstTitle.prop('aria-expanded')).toBe(true);
      expect(thirdTitle.prop('aria-expanded')).toBe(false);

      thirdTitle.simulate('click');
      expect(firstTitle.prop('aria-expanded')).toBe(false);
      expect(thirdTitle.prop('aria-expanded')).toBe(true);
    });
  });
});
