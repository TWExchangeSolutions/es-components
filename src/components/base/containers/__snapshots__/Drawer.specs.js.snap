exports[`drawer component accordion renders as expected 1`] = `
<div
  className="dtLLet">
  <div
    className="first byHzie">
    <div
      aria-expanded={true}
      className="elnnFO"
      onClick={[Function]}
      role="tab">
      <i
        aria-hidden={true}
        className="cTarNb idGniF"
        content="63d"
        fontFamily="oecom-icons"
        fontSize="inherit" />
      collapse 1
    </div>
    <div
      className="lkOJFZ"
      style={
        Object {
          "display": null,
          "height": "auto",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      }>
      <div
        className={undefined}
        style={
          Object {
            "overflow": "hidden",
          }
        }>
        first
      </div>
    </div>
  </div>
  <div
    className="second byHzie">
    <div
      aria-expanded={false}
      className="elnnFO"
      onClick={[Function]}
      role="tab">
      <i
        aria-hidden={true}
        className="cTarNb dyzqxb"
        content="63f"
        fontFamily="oecom-icons"
        fontSize="inherit" />
      collapse 2
    </div>
    <div
      className="lkOJFZ"
      style={
        Object {
          "display": "none",
          "height": "0px",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
  <div
    className="third byHzie">
    <div
      aria-expanded={false}
      className="elnnFO"
      onClick={[Function]}
      role="tab">
      <i
        aria-hidden={true}
        className="cTarNb dyzqxb"
        content="63f"
        fontFamily="oecom-icons"
        fontSize="inherit" />
      collapse 3
    </div>
    <div
      className="lkOJFZ"
      style={
        Object {
          "display": "none",
          "height": "0px",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
</div>
`;

exports[`drawer component drawer renders as expected 1`] = `
<div
  className="important dtLLet">
  <div
    className="first byHzie">
    <div
      aria-expanded={false}
      className="elnnFO"
      onClick={[Function]}
      role="tab">
      <i
        aria-hidden={true}
        className="cTarNb dyzqxb"
        content="63f"
        fontFamily="oecom-icons"
        fontSize="inherit" />
      collapse 1
    </div>
    <div
      className="lkOJFZ"
      style={
        Object {
          "display": "none",
          "height": "0px",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
  <div
    className="second byHzie">
    <div
      aria-expanded={false}
      className="elnnFO"
      onClick={[Function]}
      role="tab">
      <i
        aria-hidden={true}
        className="cTarNb dyzqxb"
        content="63f"
        fontFamily="oecom-icons"
        fontSize="inherit" />
      collapse 2
    </div>
    <div
      className="dIyrRA"
      style={
        Object {
          "display": "none",
          "height": "0px",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
  <div
    className="third byHzie">
    <div
      aria-expanded={false}
      className="elnnFO"
      onClick={[Function]}
      role="tab">
      <i
        aria-hidden={true}
        className="cTarNb dyzqxb"
        content="63f"
        fontFamily="oecom-icons"
        fontSize="inherit" />
      collapse 3
    </div>
    <div
      className="lkOJFZ"
      style={
        Object {
          "display": "none",
          "height": "0px",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
</div>
`;
